compiler/backend/src/org/jetbrains/kotlin/codegen/state/KotlinTypeMapper.kt

```diff
@@ -910,7 +910,8 @@ class KotlinTypeMapper @JvmOverloads constructor(
         val isConstructor = isConstructor(jvmSignature)
         val descriptor = getDefaultDescriptor(
             jvmSignature,
-            if (isStaticMethod(kind, functionDescriptor) || isConstructor) null else ownerType.descriptor,
+            if (isStaticMethod(kind, functionDescriptor) || isStaticDeclaration(functionDescriptor) || isConstructor)
```

Comment:
Please add `isIrBackend &&` here so code would be cleaned after switching to new typeMapper

Reply:
Are you sure that this isn't a bug in the current TypeMapper? The function in question will produce the wrong result if you call it with a descriptor for a static function which is not at top-level (or in an inline class) and not marked with the `@JvmStatic` annotation.

Apparently the jvm backend does not currently run into this case, but it's possible that it will with future language extensions.

Checking for static declarations (`dispatchReceiver == null`) in addition to the other special cases seems more robust to me.

Reply:
Yes, you are right: it would cause some problems in future. Thanks!

Reply:
FYI we've met some problems after bootstrapping on the compiler with this change: https://teamcity.jetbrains.com/viewLog.html?buildId=2319700&buildTypeId=Kotlin_dev_CompilerAndPlugin_191. I did not have the time yet to investigate, and the problem did not reproduce on a small sample. So meanwhile I've restored the old behavior for the non-IR backend here in 4e1effd0655eac4a971d5f23e94652d08e544774.

In general, let's please be more careful with such changes in the future. Even though we have a lot of codegen tests, they unfortunately don't cover everything, so even if the change itself is correct in theory (as in this case), interaction with code from some other dark corners of our codebase may lead to such errors. Specifically, I'd prefer if we didn't change the behavior of the old backend "speculatively", i.e. without understanding of the real cases where behavior changes.

Reply:
That's humbling. I'll be more careful about touching the old backend in the future...

Reply:
@sfs Thanks!